### ISSUE TITLE
Nullability information is seemingly lost for members of a variable declared by a pattern

### ISSUE THREAD

**Issue by PathogenDavid** [dotnet/roslyn#47676]
**Version Used**: Compiler version: '3.8.0-4.20462.1 (204bbc36)'. Language version: preview.

**Steps to Reproduce**:

Compile the following: ([SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AxAHYCuANuRMOXAARzHW0CwAUOwAIDMdnATHQCC7AN7s6kvrwBCAfjoyAYpjjkAJgG52AX3ZdeAxWIlSedYAHtL5OgBU4sOqLoBzODE109bHwb6CDrAAsh4AFpbqUCZsUtIW1rZBMACMABRCdBAAlKZxkgC8AHxZdJhQws6KKmrqIHTEljANFLY6dABkHVkAdMqqGj3J2rFSefFWNvaOMPxplsAAVnAAxs2WuaP5dMV0lmUVmRCd3RB9NRoHDU0tlCe9/bVDMyNx4+aTSTPc80ur65ttoUSvtypUXI8NPVGs0yHd2scug8LupnrBXkCAPSYoG4vH4gB6ROJRPGcWxdAA7hAEMRMMRXHQAMIAZQAHAA2AAM/HqABFEHBsILiCt6JZsKUAA6WKBQTA0XC3WwIIUisU9cbvXifaawAAsv2Waz2gKk4i2+RwdHmVyOZu2FvxkmtaWOYIhKOhNzhbQdeKdzriR2OBSyGKDkk4AHYsmdIajhmSgT5cX5LXEY3RsBByFA4BHvLp2EA==))

```csharp
#nullable enable

public class A
{
    public B? BField;
}

public class B
{
    public bool Test { get; }
}

public class TestMethods
{
    public bool Test1(A a)
        => a is A { BField: not null } && a.BField.Test;
    
    public bool Test2(object o)
        => o is A a && a.BField is not null && a.BField.Test;
    
    public bool Test3(object o)
        => o is A { BField: not null } a && a.BField.Test;
        //                                  ^^^^^^^^
        // warning CS8602: Dereference of a possibly null reference.

    public bool Test4(object o)
    {
        if (o is A a)
        {
            if (a is { BField: not null })
            {
                A aa = a;
                return aa.BField.Test;
            }
        }

        return false;
    }
}
}
```

**Expected Behavior**: Compiles without any warnings, especially since `Test2` and `Test3` have identical codegen.

**Actual Behavior**: `Test3` has the warning noted in the comment.

-----

Originally I thought this was purely a limitation in the flow analysis, but the fact that `Test1` and `Test4` work while `Test3` didn't stuck me as odd. It seems like the nullability information of `(o as A).BField` doesn't carry over to `a.BField`.

**Comment by jcouv** [comment]
Verified this is fixed by https://github.com/dotnet/roslyn/pull/51001 (merged for 16.10)
I'll close the issue. Thanks

**Comment by PathogenDavid** [comment]
Thanks a lot!
